### PR TITLE
feat(linter): implement `no-undef`

### LINKED ISSUE #619: feat(linter): eslint/no-undef
Implement ESLint's [`no-undef`])[https://eslint.org/docs/latest/rules/no-undef#rule-details] rule.

To get started, run
```sh
just new-rule no-undef
```

### PR DESCRIPTION
**Pull request by makotot**
Closes https://github.com/web-infra-dev/oxc/issues/619

I have tried to implement https://eslint.org/docs/latest/rules/no-undef

Note: Unsupported items are excluded from test cases in this PR by commenting them out..

### CODE REVIEW

File: crates/oxc_linter/src/rules/eslint/no_undef.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            .get(0)
+            .and_then(|config| config.get("typeof"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or_default();
+        Self { type_of }
+    }
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::IdentifierReference(ident) = node.kind() {
+            if ctx.semantic().is_reference_to_global_variable(ident)
+                && !BUILTINS.contains_key(ident.name.as_str())
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_undef.rs:53]: Do we plan on supporting `env`s?

File: crates/oxc_linter/src/rules/eslint/no_undef.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let type_of = value
+            .get(0)
+            .and_then(|config| config.get("typeof"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or_default();
+        Self { type_of }
+    }
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::IdentifierReference(ident) = node.kind() {
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/no_undef.rs:51]: @Boshen what should we do about type symbols? I don't think we bind any of typescript's `lib.d.ts` files to the top-level scope

File: crates/oxc_linter/src/rules/eslint/no_undef.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/eslint/no_undef.rs:53]: Once we start reading eslint config.

File: crates/oxc_linter/src/rules/eslint/no_undef.rs (hunk 2 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/eslint/no_undef.rs:51]: We don't lint `.d.ts` files.

**Review by Boshen — CHANGES_REQUESTED** [review]
Great work! I think we can make this performant by using `run_once`, and then loop the global unreferenced variables and print their references. Would you like to give this a try?

File: crates/oxc_linter/src/rules/eslint/no_undef.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    })
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+
+    let pass = vec![
+        ("var a = 1, b = 2; a;", None),
+        // ("/*global b*/ function f() { b; }", None),
```
Comment (mysteryven) [inline crates/oxc_linter/src/rules/eslint/no_undef.rs:78]: @Boshen Do we have plans about support parsing comments in the future?

Comment (Boshen) [reply crates/oxc_linter/src/rules/eslint/no_undef.rs:78]: We can add them, we already have the comments.

File: crates/oxc_linter/src/rules/eslint/no_undef.rs (hunk 2 above)
Comment (DonIsaac) [reply crates/oxc_linter/src/rules/eslint/no_undef.rs:51]: Correct, I'm talking about using them as an "environment", the same way tsc provides "built-in" types

**Comment by DonIsaac** [comment]
This looks great to me, ready for review @mysteryven?

**Comment by makotot** [comment]
Fixed to use `run_once` method.`ctx.semantic().nodes()` is still used for typeof operator checks though.
Can I ask to review it again?

**Review by mysteryven — DISMISSED** [review]

File: crates/oxc_linter/src/rules/eslint/no_undef.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        let symbol_table = ctx.symbols();
+
+        for reference_id_list in ctx.scopes().root_unresolved_references().values() {
+            for &reference_id in reference_id_list {
+                let reference = symbol_table.get_reference(reference_id);
+                if BUILTINS.contains_key(reference.name().as_str()) {
+                    return;
+                }
+                if !self.type_of {
+                    for node in ctx.semantic().nodes().iter() {
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/eslint/no_undef.rs:60]: Oh, this is not looking good.

We need to add `AstNodeId` to `Reference`. https://github.com/web-infra-dev/oxc/issues/685

Comment (Boshen) [reply crates/oxc_linter/src/rules/eslint/no_undef.rs:60]: @makotot You may try adding `AstNodeId` to `Reference`, should be trivial (maybe).

Comment (makotot) [reply crates/oxc_linter/src/rules/eslint/no_undef.rs:60]: Thanks! I'll take a look.

Comment (makotot) [reply crates/oxc_linter/src/rules/eslint/no_undef.rs]: Fixed by https://github.com/web-infra-dev/oxc/pull/672/commits/5d16568bce3d1ab1736c26776dfdac0f8372da7b.

**Comment by makotot** [comment]
I think it is ready for review again, so I'll remove the draft status.

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Nice work! I hope you enjoyed the long journey 😁

**Pull request merged**